### ISSUE TITLE
automated builds run very slow and fail often

### ISSUE THREAD

**Issue by teodorescuserban** [docker/hub-feedback#468]
Hi,

Since a few weeks, the automated builds I use start to have problems.

During US timezone bussiness hours the build would work extremely slow and often timeout (see below). Outside this interval build would error out quite often anyway (usually they stay in ```claimed``` state for a while then timeout or throw another error).

![docker_hub_timeouts](https://cloud.githubusercontent.com/assets/1542984/11206321/ee3fa04c-8d16-11e5-97fd-3f5e1793a38f.png)

Is there any solution?

If this continue I will be forced to find another solution (maybe just integrating the Dockerfile files in a github repo and just build the images locally on each server - seemed like a stupid thing to do at first, but I cant wait 4 hours for a simple image build).

Thank you!

**Comment by jfisbein** [comment]
I'm having the same problem.

```json
Push failed: {"errorDetail":{"message":"Head https://dseasb33srnrn.cloudfront.net/registry-v2/docker/registry/v2/blobs/sha256/23/2375d467c689c2532df1204708299403e0121a5928e4184a5663a2779cab14dc/data?Expires=1447758071\u0026Signature=JqLiwjL8cu6vldd4ns4O7c506k1JzkvYg6jg6CGPb4xyXQ5XJYCdfedY9E7NMMfXbyMza0OqwQ3v9E4WEIBGPgStFrwWCYQka4rqYq67XuQtP3FCaAflr-Q4t9BBQjBs3Wvdg3ceHupypnKF44wWQqMRq~EfExdCZQsDaFZhtr0_\u0026Key-Pair-Id=APKAJECH5M7VWIS5YZ6Q: dial tcp 54.230.103.20:443: i/o timeout"},"error":"Head https://dseasb33srnrn.cloudfront.net/registry-v2/docker/registry/v2/blobs/sha256/23/2375d467c689c2532df1204708299403e0121a5928e4184a5663a2779cab14dc/data?Expires=1447758071\u0026Signature=JqLiwjL8cu6vldd4ns4O7c506k1JzkvYg6jg6CGPb4xyXQ5XJYCdfedY9E7NMMfXbyMza0OqwQ3v9E4WEIBGPgStFrwWCYQka4rqYq67XuQtP3FCaAflr-Q4t9BBQjBs3Wvdg3ceHupypnKF44wWQqMRq~EfExdCZQsDaFZhtr0_\u0026Key-Pair-Id=APKAJECH5M7VWIS5YZ6Q: dial tcp 54.230.103.20:443: i/o timeout"}
```

**Comment by makuk66** [comment]
See https://github.com/docker/hub-feedback/issues/466

**Comment by christoph-hipp** [comment]
Dear Sirs,

since recently we are encountering errors in automated build repositories. The build completes, when the image gets pushed to the repository, the push seems to time out:

Push failed: {"errorDetail":{"message":"Head https://dseasb33srnrn.cloudfront.net/registry-v2/docker/registry/v2/blobs/sha256/1d/1dfefe48b93b2e79427899e2141b53b91421c3c7c67487ddb87f5341998ea2bc/data?Expires=1447767419\u0026Signature=ECtahpcokJ4nHId1shHYulOsJXy4kBPmmY77qUGdlDq8mDao5KlcOUuugx5FORNesw6gbOhgjU4vNFUgMmw8KWW9slENyOKkGNF-r0pZhO9VzW9YN95jaLR0U8Ynh6GK-CLANnXOmiQroHjHPjIozWxxdqltg6bPmaAPPHKTWz4_\u0026Key-Pair-Id=APKAJECH5M7VWIS5YZ6Q: dial tcp 54.240.160.205:443: i/o timeout"},"error":"Head https://dseasb33srnrn.cloudfront.net/registry-v2/docker/registry/v2/blobs/sha256/1d/1dfefe48b93b2e79427899e2141b53b91421c3c7c67487ddb87f5341998ea2bc/data?Expires=1447767419\u0026Signature=ECtahpcokJ4nHId1shHYulOsJXy4kBPmmY77qUGdlDq8mDao5KlcOUuugx5FORNesw6gbOhgjU4vNFUgMmw8KWW9slENyOKkGNF-r0pZhO9VzW9YN95jaLR0U8Ynh6GK-CLANnXOmiQroHjHPjIozWxxdqltg6bPmaAPPHKTWz4_\u0026Key-Pair-Id=APKAJECH5M7VWIS5YZ6Q: dial tcp 54.240.160.205:443: i/o timeout”}

This happens with most of the automated builds, even though one or two seems to complete now and then.

We have repos that are failing constantly since a day. We can give you details per email if required.

Since the issue prevents us to deploy new software, there is some urgency to it. Please advise how this can be resolved.

**Comment by koszikot** [comment]
Same thing happens to me :|

**Comment by teodorescuserban** [comment]
Apparently it is now fixed.
(at least it works better for me).


În Mie, 18 nov. 2015 la 16:20, koszikot <notifications@github.com> a scris:

**Comment by jakirkham** [comment]
Yeah, was seeing this and also believe it is fixed.

**Comment by teodorescuserban** [comment]
Well, i take it back.. 3 autobuild error out with "log unrecoverable"...

Mie, 18 nov. 2015, 22:12, jakirkham <notifications@github.com> a scris:

**Comment by jakirkham** [comment]
@teodorescuserban, the "Log unrecoverable" is an unrelated issue ( https://github.com/docker/hub-feedback/issues/417 ). In order to fix these terrible build issues, they reverted recent changes that fixed that bug ( https://github.com/docker/hub-feedback/issues/417#issuecomment-158471416 ). Personally, I would rather see "Log unrecoverable" than issues like this one.

**Comment by teodorescuserban** [comment]
Thank you!

Sâm, 21 nov. 2015, 20:47, jakirkham <notifications@github.com> a scris:

**Comment by sheerun** [comment]
I just had similar issue (build is stuck on "claimed"). Build id: bcpsnhpkyuntfrlrcqtnknb

**Comment by schmunk42** [comment]
Several of my automated builds also get queued for over an hour.

I know that this is a free service, but I also need to say that waiting over an hour renders the automated builds very ineffective.

**Comment by jakirkham** [comment]
I think this is unrelated @schmunk42. It looks like they updated the Ubuntu image. This results in a huge load on the server due to the number of dependent images that need to be updated. Hence the long queue times.

**Comment by schmunk42** [comment]
@jakirkham Thank you for the info, I remember seeing long build times before, especially when `ubuntu` or `debian` got updated.

If this is the case my suggestion would be to prioritize "build-on-push" over "build-triggered-because-a-remote-got-updated".

**Comment by thaJeztah** [comment]
> If this is the case my suggestion would be to prioritize "build-on-push" over "build-triggered-because-a-remote-got-updated".

Not sure if that's easily achievable, but I like the idea

**Comment by jakirkham** [comment]
Though I do think the original issue of automated builds running slow is back.

**Comment by davidjbrewer** [comment]
It was building in less than 10 minutes for the last two (2) weeks, but in the last few days it has started to take longer.
Being a paying customer doesn't make any difference, which is wrong. SLAs should be adjusted accordingly.

**Comment by davidjbrewer** [comment]
This cloud-based service is beginning to border on unusable and about ready for the waste bin.

**Comment by HariSekhon** [comment]
I'm also very frustrated with the speed of automated builds - I think they just need more capacity to run builds?

**Comment by abobwhite** [comment]
@HariSekhon I think you're right. My single page apps (angular, ember) are taking 10+ minutes to build (npm install, bower install, build). And this is after I made a base image to download node, npm, and bower as a starting point for my app. This process takes only 1-2 minutes in my development environment.

**Comment by pkennedyr** [comment]
@HariSekhon, @abobwhite: Thanks for the feedback. We are actively working on features that will enable users to specify the size of machine on which to run builds. We will also be introducing improvements / optimizations to the core build service that will help reduce wait time as well. Stay tuned ...

**Comment by abobwhite** [comment]
@pkennedyr Thanks for the quick follow-up. I'm assuming with this added benefit there will be some sort of added cost? Do you have any info around what that might be?

**Comment by pkennedyr** [comment]
@abobwhite We'll be announcing specifics on the feature and packaging later this year ...

**Comment by jippi** [comment]
@pkennedyr still due for 2016? :)

**Comment by gmistick** [comment]
Is this still relevant?
The build is so slow!

**Comment by davidjbrewer** [comment]
Yes. It is still common to see 15+ minute builds for our larger images. We would prefer to see blindingly fast builds on Docker Hub to discourage our developers from reverting to too much client-side resources.

**Comment by abobwhite** [comment]
This problem has gotten even worse. A few months ago, all my client-side image builds jumped from about 15-20 minutes to 40+ minutes.

**Comment by pkennedyr** [comment]
For projects that have many dependencies which don’t change often, we recommend utilizing the build caching feature in [Docker Cloud](https://cloud.docker.com/) to help speed build time. 

Note: at Docker Cloud, you will also find a number of other build features (not found in Docker Hub), including: multiple builder sizes, testing on PR, streaming logs, specifying build context, and more.

For more information, please check out:
* Docker Cloud Automated Builds Docs: https://docs.docker.com/docker-cloud/builds/automated-build/
* Future Updates on Docker Hub & Docker Cloud: https://forums.docker.com/t/future-updates-on-docker-hub-docker-cloud/20498

**Comment by jakirkham** [comment]
> Is this still relevant?

I don't think this issue is the cause, but yes builds on Docker Hub are quite slow still. Is there any work being done to improve this speed?

FWIW I have builds doing little more than pulling an image that is on the order of few GBs and am seeing these builds go on ~10-~20mins easily. I have trouble believing Docker Hub has such poor bandwidth to be the cause of this problem, but can't figure out what else could be slowing this down. Maybe setting up a similar test and exploring more detailed logs (that I don't have access to) will reveal the underlying problem.

**Comment by dbroadhurst** [comment]
I'm curious about the 'build on my own nodes option'. Clicking doesn't seem to reveal much info but if there was a way to provide my own faster node to perform this automated build step it would be much better than waiting for builds that take 10x more time than on my local machine. Normally I'd solve this with my own CI/CD tools but its overkill for my personal projects

**Comment by abobwhite** [comment]
@dbroadhurst After much tinkering the last few months, I have found using that "build on my organization's nodes" (or "own" if not an org, presumably) works well! If you have a decent node integrated with docker cloud, the builds are MUCH faster. TIP: If you want to build on a specific node, you'll want to tag that node with "builder" (that part was not well documented).

**Comment by darkpixel** [comment]
I just got a paid account because I thought it would be less hassle than managing my own repo for the 7 images I need to deploy.

Local build times on a 4-core Intel(R) Core(TM) i7-3630QM CPU @ 2.40GHz take about 45 seconds for my 'biggest' package.  After loading my system down with a bunch of disk and CPU I/O and getting the load up to ~10.3 and then trying to build 7 packages at the same time, it took less than 5 minutes.

On Docker Hub my packages take anywhere from 10 to 18 minutes to get built.

I know $12/month is a pittance, but I can get packages built in under a minute with a $10/month digital ocean server...

**Comment by HariSekhon** [comment]
.

**Comment by darkpixel** [comment]
The reason there's no activity is that it has been adequately reported and documented but no one has fixed the issue yet.
Either that, or everyone has bailed off docker hub and moved to something with decent performance like the GitHub Container Registry.

**Comment by thaJeztah** [comment]
For those following this ticket, I recommend to upvote the public roadmap issue to help getting work on this prioritised; https://github.com/docker/roadmap/issues/73.

If you have specific details about what parts are slow (scheduling, actual builds), and/or failures; feel free to leave a comment there as well.


(disclaimer: I work at Docker, but I'm not on the team doing the auto builds, but hope that helps)
